KakapoExample/KakapoTests/JSONAPITests.swift

```diff
@@ -0,0 +1,231 @@
+                }
+                
+                expect(objects.count).to(equal(2))
+            }
+            
+            it("should only serialzie actual attributes into attributes") {
+                let lonelyMax = User(id: "11", name: "Max", dog: dog, cats: [])
+                let object = json(lonelyMax)
+                let attributes = object["attributes"].dictionaryValue
+                expect(attributes.count).to(equal(1)) // only name should be here, no id, no cats
+                expect(attributes["name"]).to(equal("Max"))
+                expect(attributes["cats"]).to(beNil())
+                expect(attributes["id"]).to(beNil())
+            }
+            
+            it("should fail to serialize CustomSerializable entities") {
+                // TODO: discuss because this might be unexpected
+                let object = json(CustomPost(id: "123", title: "Test"))
+                expect(object["id"].string).toNot(beNil())
+                expect(object["foo"].string).to(beNil())
+            }
+        }
+        
+        describe("JSON API Entity relationship serialization") {
+            it("should serialzie the relationships when they are single JSONAPIEntities") {
+                let object = json(user)
+                let dog = object["relationships"]["dog"]["data"]
+                expect(dog.dictionary).toNot(beNil())
+                expect(dog["id"].stringValue).to(equal("22"))
+                expect(dog["type"].stringValue).to(equal("dog"))
+            }
+            
+            it("should serialzie the relationships when they are arrays of JSONAPIEntities") {
+                let object = json(user)
+                let cats = object["relationships"]["cats"]["data"].array!
+                expect(cats.count).to(equal(2))
+                expect(cats[0]["id"].stringValue).to(equal("33"))
+                expect(cats[0]["type"].stringValue).to(equal("cat"))
+                expect(cats[1]["id"].stringValue).to(equal("44"))
+                expect(cats[1]["type"].stringValue).to(equal("cat"))
+            }
+            
+            it("should not serialzie relationships of relationships") {
+                let object = json(user)
+                let dogData = object["relationships"]["dog"]["data"].dictionary
+                expect(dogData).toNot(beNil())
+                expect(dogData?["relationships"]).to(beNil())
+            }
+            
+            it("should not serialzie attributes of relationships") {
+                let object = json(user)
+                let dogData = object["relationships"]["dog"]["data"].dictionary
+                expect(dogData).toNot(beNil())
+                expect(dogData?["attributes"]).to(beNil())
+            }
+            
+            it("should not serialzie nil relationships") {
+                let object = json(dog)
+                let cat = object["relationships"].dictionaryValue
```

Comment:
Serialization of `Dog(id: "22", name: "Joan", chasingCat: Cat(id: "33", name: "Stancho"))` adds cat as an attribute instead of a relationship because the mirroring check fails. 

Should we also handle optionals there?